### ISSUE TITLE
Problem with sort

### ISSUE THREAD

**Issue by ghost** [zeeraw/gradient#3]
The `Gradient::Map` initialiser sorts the initial set of points by location, which is fine if there are no repeated locations.  However, in order to correctly represent discontinuities in gradients, duplicate locations need to be handled.  Consider the following simple step gradient

```
[0.0, black], [0.5, black], [0.5. white], [1.0, white]
```

If we sort this array by location with a  [non-stable sort](https://en.wikipedia.org/wiki/Sorting_algorithm#Stability), such as [Ruby's](http://stackoverflow.com/questions/15442298/), then the result may be a sawtooth

```
[0.0, black], [0.5, white], [0.5. black], [1.0, white]
```

Several solutions are possible
- implement a stable sort on top of an unstable one (see the second link above)
- check if the input is sorted, if it is then do not sort it
- forbid discontinuities in the input

(those would be in my order of preference), and probably others.

**Comment by zeevallin** [comment]
I really don't like straight down forbidding certain inputs since this library is already primarily dealing with potentially dodgy inputs from the outside world. It's not very nice behaviour.

My inclination is to implement a stable sort algorithm for retaining the relative positions of the colours and have it replace the current sort in the map initialiser. Let's call that **Option A**.

Since I don't want to make the object initialiser too complex, another inclination would be to create a point sorter object which could be used to give a map expecting more strict input. Let's call that **Option B**.

A compromise between these two options would be that we have a point sorter object and invoke it on map initialisation.

Thoughts?

**Comment by ghost** [comment]
I'd prefer A, stable sort is only one line of code and runs in _O(n log n)_ time

**Comment by zeevallin** [comment]
@jjgreen Oh yeah, didn't read through that SO properly before. That really is not a lot of code. Let's add that for `0.5.1`.

**Comment by zeevallin** [comment]
Fixed in https://github.com/zeeraw/gradient/commit/25fc9017d5e8bbe9bf903f7bbf7e194185525d70 @jjgreen

**Comment by zeevallin** [comment]
Released in [`v0.5.1`](https://github.com/zeeraw/gradient/releases/tag/v0.5.1)
